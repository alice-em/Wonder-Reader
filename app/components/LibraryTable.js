import PropTypes from 'prop-types';
import React, { Component } from 'react';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableHead from '@material-ui/core/TableHead';

import LibraryItem from './LibraryItem';
import LibraryTableHead from './LibraryTableHead';

class LibraryTable extends Component {
  renderLibraryItem = content => {
    const {
      basename,
      dirname,
      id,
      isDirectory
    } = content;
    const { onContentClick } = this.props;

    return (
      <LibraryItem
        basename={basename}
        dirname={dirname}
        id={id}
        isDirectory={isDirectory}
        onRowClick={() => { onContentClick(content);}}
        style={{
          marginLeft: '5vw',
          maxHeight: '50vh',
          maxWidth: '50vw'
        }}
      />
    );
  };

  renderLibraryItems = (contents) => contents.map(this.renderLibraryItem);

  renderTableBody = (contents) => (
    <TableBody>
      {this.renderLibraryItems(contents)}
    </TableBody>
  );

  render() {
    const { contents } = this.props;
    return (
      <Table
        className="library-menu"
        selectable="false"
      >
        <TableHead>
          <LibraryTableHead />
        </TableHead>
        {this.renderTableBody(contents)}
      </Table>
    );
  }
}

LibraryTable.propTypes = {
  contents: PropTypes.arrayOf(PropTypes.object).isRequired,
  onContentClick: PropTypes.func.isRequired
};

export default LibraryTable;
