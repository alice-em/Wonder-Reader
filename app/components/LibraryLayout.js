import IconButton from '@material-ui/core/IconButton';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { FaClose, FaFolderOpen, FaLevelUp } from 'react-icons/lib/fa';

import LibraryHeader from './LibraryHeader';
import LibraryTable from './LibraryTable';

const { copyDeepObject } = require('../modules/copyData.js');
const { dialog } = require('electron').remote;
const { generateNestedContentFromFilepath } = require('../modules/generate.js');

const styles = {
  closeButton: {
    background: '#ef5350'
  },
  libraryStyles: {
    marginTop: '64px',
    maxHeight: 'calc(90vh - 64px)',
    overflowY: 'auto'
  }
};

class LibraryLayout extends Component {
  state = {
    id: 'libraryRoot',
    basename: '',
    bookmark: '',
    dirname: '',
    fullpath: null,
    isDirectory: true,
    root: '',
    contents: []
  };

  componentDidMount() {
    const { root } = this.props;
    if (root) {
      this.updateContent(root);
    }
  }

  componentWillUnmount() {
    this.props.saveContentDataToParent(this.state);
  }

  renderButtons = () => (
    <div>
      {this.renderFolderOpen()}
      {this.renderLevelUp()}
      {this.renderClose()}
    </div>
  );

  renderClose = () => (
    <IconButton
      onClick={this.props.closeLibrary}
      color="primary"
      style={styles.closeButton}
    >
      <FaClose />
    </IconButton>
  );

  renderFolderOpen = () => (
    <IconButton onClick={this.openDirectory} color="primary">
      <FaFolderOpen />
    </IconButton>
  );

  renderLevelUp = () => (
    <IconButton onClick={this.setParentAsLibrary} color="primary">
      <FaLevelUp />
    </IconButton>
  );

  renderLibary = () => {
    const { basename, bookmark, contents, dirname, fullpath, id } = this.state;

    return (
      <LibraryTable
        key={id}
        basename={basename}
        bookmark={bookmark}
        dirname={dirname}
        fullpath={fullpath}
        isDirectory
        contents={contents}
        onContentClick={this.onClick}
        saveContentDataToParent={this.saveContentDataToParent}
        saveContentsDataToParent={this.saveContentsDataToParent}
      />
    );
  };

  onClick = content => {
    if (content.isDirectory) {
      this.onDirectoryClick(content);
    } else {
      this.onFileClick(content);
    }
  };

  onDirectoryClick = content => {
    this.updateContent(content.fullpath);
  };

  onFileClick = content => {
    this.props.openComic(content.fullpath);
  };

  // Function to open `Load` window, and pass path to generateContent, then setstate
  openDirectory = () => {
    const { updateRoot } = this.props;
    dialog.showOpenDialog(
      {
        properties: ['openDirectory']
      },
      filepaths => {
        if (Array.isArray(filepaths)) {
          const filepath = filepaths[0];
          updateRoot(filepath);
          this.updateContent(filepath);
        }
      }
    );
  };

  saveContentDataToParent = content => {
    const newContent = copyDeepObject(content);
    this.setState(newContent);
  };

  saveContentsDataToParent = contents => {
    const newContent = copyDeepObject(this.state);
    newContent.contents = contents;
    this.setState({ contents: newContent });
  };

  setParentAsLibrary = () => {
    const { dirname } = this.state;
    this.updateContent(dirname);
  };

  updateContent = fullpath => {
    generateNestedContentFromFilepath(fullpath, content => {
      const newContent = content;
      newContent.id = 'libraryRoot';
      this.setState(newContent);
    });
  };

  render() {
    const { fullpath } = this.state;

    const libraryTable = fullpath ? this.renderLibary() : null;
    return (
      <div className="library" style={styles.libraryStyles}>
        <LibraryHeader
          position="fixed"
          title="Library"
          buttons={this.renderButtons()}
          onContentClick={this.onClick}
        />
        {libraryTable}
      </div>
    );
  }
}

LibraryLayout.defaultProps = {
  root: null
};

LibraryLayout.propTypes = {
  closeLibrary: PropTypes.func.isRequired,
  openComic: PropTypes.func.isRequired,
  root: PropTypes.string,
  saveContentDataToParent: PropTypes.func.isRequired,
  updateRoot: PropTypes.func.isRequired
};

export default LibraryLayout;
