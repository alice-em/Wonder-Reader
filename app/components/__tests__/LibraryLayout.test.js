import Adapter from 'enzyme-adapter-react-16';
import electron from 'electron';
import Enzyme, { shallow } from 'enzyme';
import React from 'react';

import LibraryLayout from '../LibraryLayout';
import LibraryTable from '../LibraryTable';

const {
  generateNestedContentFromFilepath,
} = require('../../modules/generate.js');

jest.mock('electron', () => ({
  remote: {
    dialog: { showOpenDialog: jest.fn() },
  },
}));
jest.mock('../../modules/generate.js', () => ({
  generateNestedContentFromFilepath: jest.fn(),
}));

Enzyme.configure({ adapter: new Adapter() });

const props = {
  closeLibrary: jest.fn(),
  openComic: jest.fn(),
  root: '.',
  saveContentDataToParent: jest.fn(),
  updateRoot: jest.fn(),
};

const sampleContent = {
  contents: [],
  dirname: '',
  fullpath: null,
  id: 'libraryRoot',
  isDirectory: true,
  root: '',
};

describe('LibraryLayout', () => {
  it('should render', () => {
    const wrapper = shallow(<LibraryLayout {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a LibraryTable with a truthy state.fullpath', () => {
    const wrapper = shallow(<LibraryLayout {...props} />);
    wrapper.setState({ fullpath: 'sample' });
    expect(wrapper.find(LibraryTable)).toHaveLength(1);
  });

  describe('Class Functions', () => {
    describe('onClick', () => {
      const fullpath = 'sample';
      const createArguments = isDirectory => ({
        fullpath,
        isDirectory,
      });
      it('should updateContent if target.isDirectory', () => {
        const wrapper = new LibraryLayout(props);
        wrapper.updateContent = jest.fn();
        wrapper.onClick(createArguments(true));
        expect(wrapper.updateContent).toHaveBeenCalledWith(fullpath);
      });

      it('should props.openComic if !target.isDirectory', () => {
        const wrapper = new LibraryLayout(props);
        wrapper.updateContent = jest.fn();
        wrapper.onClick(createArguments(false));
        expect(props.openComic).toHaveBeenCalledWith(fullpath);
      });
    });

    describe('openDirectory', () => {
      it('should show dialog', () => {
        const wrapper = new LibraryLayout(props);
        wrapper.openDirectory();
        expect(electron.remote.dialog.showOpenDialog).toHaveBeenCalledWith(
          { properties: ['openDirectory'] },
          wrapper.updateRoot,
        );
      });
    });

    describe('setContentToState', () => {
      it('sets state to content argument', (done) => {
        const wrapper = new LibraryLayout(props);
        wrapper.setState = jest.fn();
        wrapper.setContentToState(sampleContent);
        expect(wrapper.setState).toHaveBeenCalledWith(sampleContent);
        done();
      });
    });

    describe('setParentAsLibrary', () => {
      it('sets state based on dirname state', (done) => {
        const wrapper = new LibraryLayout(props);
        wrapper.setParentAsLibrary();
        expect(wrapper.state).toEqual(sampleContent);
        done();
      });
    });

    describe('updateContent', () => {
      it('should generateNestedContent and set that data to state', () => {
        const wrapper = new LibraryLayout(props);
        wrapper.setContentToState = jest.fn();
        wrapper.updateContent('.');
        expect(generateNestedContentFromFilepath).toHaveBeenCalledWith(
          '.',
          wrapper.setContentToState,
        );
      });
    });

    describe('updateRoot', () => {
      it('should updateContent with a valid [filepath]', () => {
        const sampleFilepath = 'sample';
        const wrapper = new LibraryLayout(props);
        wrapper.updateContent = jest.fn();
        wrapper.updateRoot([sampleFilepath]);
        expect(wrapper.updateContent).toHaveBeenCalledWith(sampleFilepath);
      });

      it('should not updateContent with a invalid [filepath]', () => {
        const sampleFilepath = '';
        const wrapper = new LibraryLayout(props);
        wrapper.updateContent = jest.fn();
        wrapper.updateRoot([sampleFilepath]);
        expect(wrapper.updateContent).not.toHaveBeenCalled();
      });
    });
  });
});
