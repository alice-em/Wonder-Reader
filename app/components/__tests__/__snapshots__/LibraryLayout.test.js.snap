// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LibraryLayout should render 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LibraryLayout
    closeLibrary={[MockFunction]}
    openComic={[MockFunction]}
    root="."
    saveContentDataToParent={[MockFunction]}
    updateRoot={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <LibraryHeader
          onContentClick={[Function]}
          position="fixed"
          title="Library"
        >
          <div>
            <WithStyles(IconButton)
              color="primary"
              onClick={[Function]}
            >
              <FaFolderOpen />
            </WithStyles(IconButton)>
            <WithStyles(IconButton)
              color="primary"
              onClick={[Function]}
            >
              <FaLevelUp />
            </WithStyles(IconButton)>
            <WithStyles(IconButton)
              color="primary"
              onClick={[MockFunction]}
              style={
                Object {
                  "background": "#ef5350",
                }
              }
            >
              <FaClose />
            </WithStyles(IconButton)>
          </div>
        </LibraryHeader>,
        null,
      ],
      "className": "library",
      "style": Object {
        "marginTop": "64px",
        "maxHeight": "calc(90vh - 64px)",
        "overflowY": "auto",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <div>
            <WithStyles(IconButton)
              color="primary"
              onClick={[Function]}
            >
              <FaFolderOpen />
            </WithStyles(IconButton)>
            <WithStyles(IconButton)
              color="primary"
              onClick={[Function]}
            >
              <FaLevelUp />
            </WithStyles(IconButton)>
            <WithStyles(IconButton)
              color="primary"
              onClick={[MockFunction]}
              style={
                Object {
                  "background": "#ef5350",
                }
              }
            >
              <FaClose />
            </WithStyles(IconButton)>
          </div>,
          "onContentClick": [Function],
          "position": "fixed",
          "title": "Library",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <WithStyles(IconButton)
                color="primary"
                onClick={[Function]}
              >
                <FaFolderOpen />
              </WithStyles(IconButton)>,
              <WithStyles(IconButton)
                color="primary"
                onClick={[Function]}
              >
                <FaLevelUp />
              </WithStyles(IconButton)>,
              <WithStyles(IconButton)
                color="primary"
                onClick={[MockFunction]}
                style={
                  Object {
                    "background": "#ef5350",
                  }
                }
              >
                <FaClose />
              </WithStyles(IconButton)>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <FaFolderOpen />,
                "color": "primary",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <FaLevelUp />,
                "color": "primary",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <FaClose />,
                "color": "primary",
                "onClick": [MockFunction],
                "style": Object {
                  "background": "#ef5350",
                },
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
          ],
          "type": "div",
        },
        "type": [Function],
      },
      null,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <LibraryHeader
            onContentClick={[Function]}
            position="fixed"
            title="Library"
          >
            <div>
              <WithStyles(IconButton)
                color="primary"
                onClick={[Function]}
              >
                <FaFolderOpen />
              </WithStyles(IconButton)>
              <WithStyles(IconButton)
                color="primary"
                onClick={[Function]}
              >
                <FaLevelUp />
              </WithStyles(IconButton)>
              <WithStyles(IconButton)
                color="primary"
                onClick={[MockFunction]}
                style={
                  Object {
                    "background": "#ef5350",
                  }
                }
              >
                <FaClose />
              </WithStyles(IconButton)>
            </div>
          </LibraryHeader>,
          null,
        ],
        "className": "library",
        "style": Object {
          "marginTop": "64px",
          "maxHeight": "calc(90vh - 64px)",
          "overflowY": "auto",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <div>
              <WithStyles(IconButton)
                color="primary"
                onClick={[Function]}
              >
                <FaFolderOpen />
              </WithStyles(IconButton)>
              <WithStyles(IconButton)
                color="primary"
                onClick={[Function]}
              >
                <FaLevelUp />
              </WithStyles(IconButton)>
              <WithStyles(IconButton)
                color="primary"
                onClick={[MockFunction]}
                style={
                  Object {
                    "background": "#ef5350",
                  }
                }
              >
                <FaClose />
              </WithStyles(IconButton)>
            </div>,
            "onContentClick": [Function],
            "position": "fixed",
            "title": "Library",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <WithStyles(IconButton)
                  color="primary"
                  onClick={[Function]}
                >
                  <FaFolderOpen />
                </WithStyles(IconButton)>,
                <WithStyles(IconButton)
                  color="primary"
                  onClick={[Function]}
                >
                  <FaLevelUp />
                </WithStyles(IconButton)>,
                <WithStyles(IconButton)
                  color="primary"
                  onClick={[MockFunction]}
                  style={
                    Object {
                      "background": "#ef5350",
                    }
                  }
                >
                  <FaClose />
                </WithStyles(IconButton)>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": <FaFolderOpen />,
                  "color": "primary",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {},
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": <FaLevelUp />,
                  "color": "primary",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {},
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": <FaClose />,
                  "color": "primary",
                  "onClick": [MockFunction],
                  "style": Object {
                    "background": "#ef5350",
                  },
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {},
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": [Function],
              },
            ],
            "type": "div",
          },
          "type": [Function],
        },
        null,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
